### ISSUE TITLE
Control output on empty virtual properties values

### ISSUE THREAD

**Issue by dysan54** [rfoltyns/log4j2-elasticsearch#48]
Hi,

Great work Rafal !

In order to conform to ecs i use a lot of virtual properties with thread context map, with an empty String for default, like for ex `<VirtualProperty name="event.category" value="$${ctx:event.category:-}" dynamic="true" />` .

Is there a way to filter out empty properties from output ?

Thx

**Comment by rfoltyns** [comment]
No, it's not possible. But, if you're not using Markers yet, you can achieve that with mixins:

One for LogEvent to stop ignoring Marker field:
```java
@JsonSerialize(as = LogEvent.class)
public abstract class CustomLogEventMixIn extends LogEventJacksonJsonMixIn {

    @JsonUnwrapped
    @JsonProperty(value = "event.category")
    @JsonIgnore(false)
    @JsonInclude(value = JsonInclude.Include.NON_NULL)
    @Override
    public abstract Marker getMarker();

}
```

Second one for Marker (package-private):
```java
package org.apache.logging.log4j.core.jackson;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonValue;
import org.apache.logging.log4j.Marker;

public abstract class CustomMarkerMixIn extends MarkerMixIn {

    CustomMarkerMixIn() {
        super(null);
    }

    @JsonValue
    @Override
    public abstract String getName();

    @JsonIgnore
    @Override
    public abstract Marker[] getParents();

}
```

`JacksonJsonLayout` config:
```xml
    <JacksonMixIn mixInClass="foo.bar.CustomLogEventMixIn"
                             targetClass="org.apache.logging.log4j.core.LogEvent" />
    <JacksonMixIn mixInClass="org.apache.logging.log4j.core.jackson.CustomMarkerMixIn"
                             targetClass="org.apache.logging.log4j.Marker" />
```

and then simply:
```java
Marker category = MarkerManager.getMarker("custom category");
logger.info(category, "Hello, World!");
```

Do you think it could be an acceptable solution?

**Comment by dysan54** [comment]
Yes, it could do the trick for event.category but not for the others.

I would imagined an additional attribute in virtual properties that can control generation of json output, maybe like this in `VirtualPropertiesWriter` :

```
  @Override
  public void serializeAsField(Object bean, JsonGenerator gen, SerializerProvider prov) throws Exception {
    if (virtualProperties.length > 0) {
      for (int i = 0; i < virtualProperties.length; i++) {
        String val = valueResolver.resolve(virtualProperties[i]);
        if (!val.isEmpty()) {
          gen.writeFieldName(virtualProperties[i].getName());
          gen.writeString(val);
        }
      }
    }
  }

```

**Comment by rfoltyns** [comment]
Yeah, it will work, but I need to think it through. Initially I didn't want to pack too many flags there. 

Ideally, value based inclusion should be done with Jackson's `Include`, but it's not applicable to this field.

Maybe I'll just add a configurable filters as an alternative so you can choose what you'd like to do (including writing your own ones). What do you think?

```java
String val = valueResolver.resolve(virtualProperties[i]);
if (!filter.test(val)) {
    return;
}
gen.writeFieldName(virtualProperties[i].getName());
gen.writeString(val);
```

**Comment by dysan54** [comment]
Yes it would work because in my case i don't have to apply different rules between virtual properties, just a general rule to filter blank values. 
Choosing virtual properties to which apply this filter would be a plus.

**Comment by rfoltyns** [comment]
Ok, I'll add the configurable filters and `NonBlankFilter` it in 1.4.3.

Choosing properties might be useful, but it should be possible to achieve with custom filters and default values. It might look a bit awkward, but actually allows maximum functionality with minimum amount of changes to `VirtualProperty` and the writer.
```xml
<JacksonJsonLayout>
    <VirtualProperty name="field1" value="$${ctx:val1:-someDefaultValue}"/><!-- included, default behaviour -->
    <VirtualProperty name="field2" value="$${ctx:val2:-}"/><!-- excluded by NonBlankFilter -->
    <VirtualProperty name="field3" value="$${ctx:val3:-filter2}"/><!-- excluded by UserProvidedFilter2 -->
    <VirtualProperty name="field4" value="$${ctx:val4:-filter1}"/><!-- excluded by UserProvidedFilter1 -->
    <VirtualProperty name="field5" value="$${ctx:val5:-someDefaultValue}"/><!-- excluded by UserProvidedFieldFilter -->
    <VirtualProperty name="field6" value="$${ctx:val6:-someOtherValue}"/><!-- included, default behaviour -->
    <UserProvidedFilter1/>
    <UserProvidedFilter2/>
    <NonBlankFilter/><!-- available out-of-the-box -->
    <UserProvidedFieldFilter fields="field5,field6" condition="eq someDefaultValue" />
    
</JacksonJsonLayout>
```

**Comment by dysan54** [comment]
It's perfect, many thanks !

**Comment by rfoltyns** [comment]
I'll release it in 1.4.3. Enjoy! :)

**Comment by rfoltyns** [comment]
Released in 1.4.3.
